src/kubernetes_cluster/proof/builtin_controllers.rs

```diff
@@ -0,0 +1,274 @@
+    stateful_set::*,
+};
+use crate::kubernetes_cluster::spec::{
+    builtin_controllers::types::{built_in_controller_req_msg, BuiltinControllerChoice},
+    cluster::*,
+    cluster_state_machine::Step,
+    controller::common::{controller_req_msg, ControllerActionInput, ControllerStep},
+    message::*,
+};
+use crate::reconciler::spec::reconciler::Reconciler;
+use crate::temporal_logic::{defs::*, rules::*};
+use vstd::prelude::*;
+
+verus! {
+
+impl <K: ResourceView, E: ExternalAPI, R: Reconciler<K, E>> Cluster<K, E, R> {
+
+pub open spec fn every_update_msg_sets_owner_references_as(
+    key: ObjectRef, requirements: FnSpec(Option<Seq<OwnerReferenceView>>) -> bool
+) -> StatePred<Self> {
+    |s: Self| {
+        forall |msg: Message|
+            #[trigger] s.message_in_flight(msg)
+            && msg.dst.is_KubernetesAPI()
+            && msg.content.is_update_request()
+            && msg.content.get_update_request().key == key
+            ==> requirements(msg.content.get_update_request().obj.metadata.owner_references)
+    }
+}
+
+pub open spec fn every_create_msg_sets_owner_references_as(
+    key: ObjectRef, requirements: FnSpec(Option<Seq<OwnerReferenceView>>) -> bool
+) -> StatePred<Self> {
+    |s: Self| {
+        forall |msg: Message|
+            #[trigger] s.message_in_flight(msg)
+            && msg.dst.is_KubernetesAPI()
+            && msg.content.is_create_request()
+            && msg.content.get_create_request().namespace == key.namespace
+            && msg.content.get_create_request().obj.metadata.name.get_Some_0() == key.name
+            && msg.content.get_create_request().obj.kind == key.kind
+            ==> requirements(msg.content.get_create_request().obj.metadata.owner_references)
+    }
+}
+
+pub open spec fn objects_owner_references_satisfies(key: ObjectRef, requirements: FnSpec(Option<Seq<OwnerReferenceView>>) -> bool) -> StatePred<Self> {
+    |s: Self| {
+        s.resource_key_exists(key) ==> requirements(s.resource_obj_of(key).metadata.owner_references)
+    }
+}
+
+pub open spec fn objects_owner_references_violates(key: ObjectRef, requirements: FnSpec(Option<Seq<OwnerReferenceView>>) -> bool) -> StatePred<Self> {
+    |s: Self| {
+        s.resource_key_exists(key) && !requirements(s.resource_obj_of(key).metadata.owner_references)
+    }
+}
+
+pub open spec fn object_has_no_finalizers(key: ObjectRef) -> StatePred<Self> {
+    |s: Self| {
+        s.resource_key_exists(key)
+        ==> s.resource_obj_of(key).metadata.finalizers.is_None()
+    }
+}
+
+spec fn exists_delete_request_msg_in_flight_with_key(key: ObjectRef) -> StatePred<Self> {
+    |s: Self| {
+        exists |msg: Message| {
+            #[trigger] s.message_in_flight(msg)
+            && msg.dst.is_KubernetesAPI()
+            && msg.content.is_delete_request_with_key(key)
+        }
+    }
+}
+/// The proof of spec |= true ~> objects_owner_references_satisfies(eventual_owner_ref) consists of two parts:
```

Comment:
Before explaining the proof, could you first write about how should one use this lemma (in other words, why we introduce this lemma)?